[SOUP] Add support for the Cookie Store API

TingPing opened the pull request:
#### cb077473952d7036cceedb1a721502d74ee766c5
<pre>
[SOUP] Add support for the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=286189">https://bugs.webkit.org/show_bug.cgi?id=286189</a>

Reviewed by Michael Catanzaro.

NetworkStorageSession had two methods that soup did not implement: cookiesForDOMAsVector() and setCookieFromDOM()

These methods were implemented matching macOS&apos; behavior along with some minor refactoring to ensure all cookie code-paths do the same thing.

Once these methods are implemented it depends on the HAVE(COOKIE_CHANGE_LISTENER_API) build option, this involves calling out from the
cookiesDidChange to observers managed with startListeningForCookieChangeNotifications()/stopListeningForCookieChangeNotifications().

Enabling that option then enables WebCookieCache, a cache of cookies maintained in the WebProcess, WebCookieCacheSoup.cpp just
makes an in-memory SoupCookieJar.

In order to test these changes the cookies tests in web-platform-tests were ran, these were largely just outdated so had
to be rebaselined.

* LayoutTests/imported/w3c/web-platform-tests/cookies/samesite/multiple-samesite-attributes.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_no_name_and_no_value.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_no_name_equals_in_value.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_no_name_multiple_values.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.serviceworker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any.serviceworker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_origins.sub.https-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/encoding.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/encoding.https.any.serviceworker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookieStore_cross_origin.https.sub-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/attributes-ctl.sub-expected.txt: Renamed from LayoutTests/platform/gtk/imported/w3c/web-platform-tests/cookies/attributes/attributes-ctl.sub-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/domain.sub-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/invalid-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/max-age-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/name/name-ctl-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/name/name-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.document-cookie-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.document-cookie.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.header-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.header.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.document-cookie-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.document-cookie.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.header-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.header.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/document-cookie.non-secure-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite-none-secure/cookies-without-samesite-must-be-secure.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/value/value-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/cookies/attributes/domain.sub-expected.txt: Removed.
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::NetworkStorageSession):
(WebCore::NetworkStorageSession::cookiesDidChange):
(WebCore::NetworkStorageSession::setCookieStorage):
(WebCore::NetworkStorageSession::getCredentialFromPersistentStorage):
(WebCore::NetworkStorageSession::saveCredentialToPersistentStorage):
(WebCore::NetworkStorageSession::setCookieFromDOM const):
(WebCore::NetworkStorageSession::replaceCookies):
(WebCore:: const):
(WebCore::cookiesToHeader):
(WebCore::NetworkStorageSession::getRawCookies const):
(WebCore::NetworkStorageSession::domCookiesForHost):
(WebCore::NetworkStorageSession::cookiesForDOM const):
(WebCore::NetworkStorageSession::cookiesForDOMAsVector const):
(WebCore::NetworkStorageSession::cookieRequestHeaderFieldValue const):
(WebCore::NetworkStorageSession::startListeningForCookieChangeNotifications):
(WebCore::NetworkStorageSession::stopListeningForCookieChangeNotifications):
(WebCore::cookiesForSession): Deleted.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::shouldTreatAsSameSite):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
* Source/WebKit/WebProcess/WebPage/soup/WebCookieCacheSoup.cpp: Added.
(WebKit::WebCookieCache::inMemoryStorageSession):

Canonical link: <a href="https://commits.webkit.org/289693@main">https://commits.webkit.org/289693@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d671b65e8e2cf9f5bce57058d00980a15a5af55e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87686 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7202 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42070 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92551 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38432 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7583 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15373 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67697 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25445 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90688 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79342 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48069 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5571 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33744 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37543 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80484 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75969 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34625 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94437 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86461 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14854 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10895 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76547 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15108 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75198 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75782 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18638 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20148 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18582 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7813 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14870 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20172 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108955 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14614 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26202 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18058 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16396 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/max-age-expected.txt:
```diff
@@ -4,9 +4,9 @@ PASS Set cookie with age
 PASS Expire later cookie with same name and max-age=0
 PASS Expire later cookie with same name and max-age=0, and don't set cookie with max-age=0
-PASS Set mulitiple cookies with valid max-age values
+FAIL Set mulitiple cookies with valid max-age values assert_equals: The cookie was set as expected. expected "test=\"9! = foo; test9" but got "test9; test=\"9! = foo"
```
This reordering bug seems to be affecting many of the tests.

TingPing replied:
Fun fact, `soup_cookie_jar_all_cookies()` is always in reverse order.

carlosgcampos commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -70,14 +71,53 @@ NetworkStorageSession::~NetworkStorageSession()
+            return;
+
+        auto host = String::fromUTF8(soup_cookie_get_domain(newCookie));
+        g_message("NetworkStorageSession::cookiesDidChange %s", host.utf8().data());
```
Is this a debugging leftover, or do you really want to keep this message here?

carlosgcampos commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -70,14 +71,53 @@ NetworkStorageSession::~NetworkStorageSession()
+        return;
+    }
+
+    if (!newCookie && oldCookie) {
```
Isn't newCookie always nullptr at this point? We returned before if it wasn't.

carlosgcampos commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -548,72 +620,40 @@ void NetworkStorageSession::hasCookies(const RegistrableDomain& domain, Completi
 
     auto cookieURI = sameSiteInfo.isSameSite ? urlToSoupURI(url) : nullptr;
-    GUniquePtr<GSList> cookies(soup_cookie_jar_get_cookie_list_with_same_site_info(cookieStorage(), uri.get(), firstPartyURI.get(), cookieURI.get(), TRUE, sameSiteInfo.isSafeHTTPMethod, sameSiteInfo.isTopSite));
+    GSList* cookies(soup_cookie_jar_get_cookie_list_with_same_site_info(cookieStorage(), uri.get(), firstPartyURI.get(), cookieURI.get(), forHTTPHeader == ForHTTPHeader::Yes,
```
```suggestion
    GSList* cookies = soup_cookie_jar_get_cookie_list_with_same_site_info(cookieStorage(), uri.get(), firstPartyURI.get(), cookieURI.get(), forHTTPHeader == ForHTTPHeader::Yes,
```

carlosgcampos commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -548,72 +620,40 @@ void NetworkStorageSession::hasCookies(const RegistrableDomain& domain, Completi
-    UNUSED_PARAM(firstParty);
-    UNUSED_PARAM(sameSiteInfo);
-    UNUSED_PARAM(firstParty);
+    GSList* cookies(soup_cookie_jar_get_cookie_list(cookieStorage(), uri.get(), forHTTPHeader == ForHTTPHeader::Yes));
```
```suggestion
    GSList* cookies = soup_cookie_jar_get_cookie_list(cookieStorage(), uri.get(), forHTTPHeader == ForHTTPHeader::Yes);
```

spenap commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -157,20 +158,27 @@ class NetworkStorageSession
 
+    enum class ForHTTPHeader : bool { No, Yes };
+    std::optional<GSList*> lookupCookies(const URL& firstParty, const SameSiteInfo&, const URL&, ForHTTPHeader, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ApplyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking, IncludeSecureCookies, bool* didAccessSecureCookies = nullptr) const;
+    std::pair<String, bool> lookupCookiesHeaders(const URL& firstParty, const SameSiteInfo&, const URL&, ForHTTPHeader, std::optional<FrameIdentifier>, std::optional<PageIdentifier>, ApplyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking, IncludeSecureCookies) const;
```
The argument list,other than `didAccessSecureCookies`, is shared between these two methods, and it's lengthy enough that maybe we could consider extracting them into a helper structure?

TingPing replied:
I did this originally but it didn't really do anything of value IMO.

Like the function signature is more simple `std::pair<String, bool> lookupCookiesHeaders(const CookieLookupData&);`

But the usage is *identical*:

```cpp
lookupCookiesHeaders({
        firstParty,
        sameSiteInfo,
        url,
        ForHTTPHeader::Yes,
        frameID,
        pageID,
        applyTrackingPrevention,
        relaxThirdPartyCookieBlocking,
        includeSecureCookies});
```

So I decided to avoid defining a custom type in this header since its shared between all ports, so simple is better.

Dunno if anyone else has an opinion on it.

mcatanzaro replied:
I wouldn't use a helper structure for this.

spenap commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -626,37 +666,184 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
+        rawCookies.append(Cookie(cookie.get()));
+    }
+
+    g_slist_free(*soupCookies);
```
This has been quite interesting to me -- I have learnt that `g_slist_free` won't release elements, and that works fine because the code has previously wrapped them into unique pointers that will take care of that -- and when a `Cookie` object is built, it copies things based on the raw cookie but won't keep any soon-to-be-invalid pointer.

TingPing replied:
It is unclear but `GUniquePtr<SoupCookie> cookie(static_cast<SoupCookie*>(iter->data));` frees the cookie.

So the only data we have to manually free is the list.

Slightly awkward but I wanted to avoid doing something like making a Vector and converting back to GSList for libsoup, etc.

spenap replied:
Yes, the `GUniquePtr<SoupCookie>` part was fine -- I assumed it'd do the same as `unique_ptr` and would invoke some _destroy_ function on the `SoupCookie` as it went out of scope. 

What worried me was the `explicit Cookie(SoupCookie*)` constructor, as if it were to keep that pointer, it would be invalid, but it doesn't :+1:

spenap commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -626,37 +666,184 @@ static std::pair<String, bool> cookiesForSession(const NetworkStorageSession& se
+    auto host = url.host().utf8().data();
+
+    // FIXME: libsoup needs an API so we can avoid going through all cookies.
+    GUniquePtr<GSList> soupCookies(soup_cookie_jar_all_cookies(cookieStorage()));
```
...and I see how we have specialised this for `_free` and not `full`, so again we need to wrap elements in unique pointers :+1:

TingPing replied:
Same here this uses `GUniquePtr<SoupCookie> soupCookie(static_cast<SoupCookie*>(iter->data));` below.

mcatanzaro commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -70,14 +71,52 @@ NetworkStorageSession::~NetworkStorageSession()
-void NetworkStorageSession::cookiesDidChange(NetworkStorageSession* session)
+void NetworkStorageSession::cookiesDidChange(NetworkStorageSession* session, SoupCookie* oldCookie, SoupCookie* newCookie, SoupCookieJar*)
 {
     if (session->m_cookieObserverHandler)
```
I was a little surprised to see that you don't handle oldCookie and newCookie being set at the same time. Might as well document this assumption:

```suggestion
    ASSERT(!oldCookie || !newCookie);

    if (session->m_cookieObserverHandler)
```

TingPing replied:
This is on purpose, that case is a changed cookie, which the observer still needs to know about.

mcatanzaro replied:
Your current code will call `observer.cookiesAdded` and then return without calling `observer.cookiesDeleted`. You're saying this is intentional? It looks like a bug, so might want to add a clarification.

TingPing replied:
Yes, added a comment.

mcatanzaro commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -548,72 +627,40 @@ void NetworkStorageSession::hasCookies(const RegistrableDomain& domain, Completi
-bool NetworkStorageSession::getRawCookies(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking relaxThirdPartyCookieBlocking, Vector<Cookie>& rawCookies) const
+// A GSList* is used because other libsoup APIs take this.
+// optional is used with a pointer because `nullptr` means empty list and we also want to convey failures.
+std::optional<GSList*> NetworkStorageSession::lookupCookies(const URL& firstParty, const SameSiteInfo& sameSiteInfo, const URL& url, ForHTTPHeader forHTTPHeader, std::optional<FrameIdentifier> frameID, std::optional<PageIdentifier> pageID, ApplyTrackingPrevention applyTrackingPrevention, ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking, IncludeSecureCookies includeSecureCoo ...
```
I know this is already complicated, but if you want to return a `GSList*` you should at least store ownership in a smart pointer. How about `std::optional<std::unique_ptr<GSList*, soup_cookies_free>>`? You could do `using CookieList = std::unique_ptr<GSList*, soup_cookies_free>` and use it throughout this file, then you can get rid of a bunch of manual calls to `soup_cookies_free()` and just return a `std::optional<CookieList>` here.

mcatanzaro replied:
You probably would actually need a null-safe wrapper around `soup_cookies_free()`.

TingPing replied:
Great idea, could you review the new `CookieList` usage?

Also `soup_cookies_free()`, aka `g_slist_free_full()` is `NULL` safe.

mcatanzaro commented on Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
```diff
@@ -531,136 +622,248 @@ Vector<Cookie> NetworkStorageSession::getCookies(const URL& url)
-                    cookies = g_slist_remove_link(cookies, item);
-                    item = next;
+                    auto* oldList = cookies.release();
+                    cookies = CookieList(g_slist_remove_link(oldList, item));
```
Is `item` leaked?

TingPing replied:
Yeah, always was. I assumed `g_slist_remove_link` was `g_slist_delete_link`.

mcatanzaro reviewed (approved):
Looks fine to me.

cgarcia might want to review it as well.

I also considered the name `CookieListPtr` and `CookieListUniquePtr` since I don't normally like hiding pointer types behind a typedef, but I think just `CookieList` is probably fine.

TingPing committed:
[SOUP] Add support for the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=286189

Reviewed by Michael Catanzaro.

NetworkStorageSession had two methods that soup did not implement: cookiesForDOMAsVector() and setCookieFromDOM()

These methods were implemented matching macOS' behavior along with some minor refactoring to ensure all cookie code-paths do the same thing.

Once these methods are implemented it depends on the HAVE(COOKIE_CHANGE_LISTENER_API) build option, this involves calling out from the
cookiesDidChange to observers managed with startListeningForCookieChangeNotifications()/stopListeningForCookieChangeNotifications().

Enabling that option then enables WebCookieCache, a cache of cookies maintained in the WebProcess, WebCookieCacheSoup.cpp just
makes an in-memory SoupCookieJar.

In order to test these changes the cookies tests in web-platform-tests were ran, these were largely just outdated so had
to be rebaselined.

* LayoutTests/imported/w3c/web-platform-tests/cookies/samesite/multiple-samesite-attributes.https-expected.txt:
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_http_cookie_and_set_cookie_headers.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_no_name_and_no_value.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_no_name_equals_in_value.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_no_name_multiple_values.https.window-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.serviceworker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any.serviceworker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_origins.sub.https-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/encoding.https.any-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/encoding.https.any.serviceworker-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookieStore_cross_origin.https.sub-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/attributes-ctl.sub-expected.txt: Renamed from LayoutTests/platform/gtk/imported/w3c/web-platform-tests/cookies/attributes/attributes-ctl.sub-expected.txt.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/domain.sub-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/invalid-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/attributes/max-age-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/name/name-ctl-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/name/name-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.document-cookie-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.document-cookie.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.header-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__host.header.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.document-cookie-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.document-cookie.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.header-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/__secure.header.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/prefix/document-cookie.non-secure-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/samesite-none-secure/cookies-without-samesite-must-be-secure.https-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/value/value-ctl-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/cookies/value/value-expected.txt:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/cookies/attributes/domain.sub-expected.txt: Removed.
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp:
(WebCore::NetworkStorageSession::NetworkStorageSession):
(WebCore::NetworkStorageSession::cookiesDidChange):
(WebCore::NetworkStorageSession::setCookieStorage):
(WebCore::NetworkStorageSession::getCredentialFromPersistentStorage):
(WebCore::NetworkStorageSession::saveCredentialToPersistentStorage):
(WebCore::NetworkStorageSession::setCookieFromDOM const):
(WebCore::NetworkStorageSession::replaceCookies):
(WebCore:: const):
(WebCore::cookiesToHeader):
(WebCore::NetworkStorageSession::getRawCookies const):
(WebCore::NetworkStorageSession::domCookiesForHost):
(WebCore::NetworkStorageSession::cookiesForDOM const):
(WebCore::NetworkStorageSession::cookiesForDOMAsVector const):
(WebCore::NetworkStorageSession::cookieRequestHeaderFieldValue const):
(WebCore::NetworkStorageSession::startListeningForCookieChangeNotifications):
(WebCore::NetworkStorageSession::stopListeningForCookieChangeNotifications):
(WebCore::cookiesForSession): Deleted.
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::shouldTreatAsSameSite):
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/WebProcess/WebPage/WebCookieCache.cpp:
* Source/WebKit/WebProcess/WebPage/soup/WebCookieCacheSoup.cpp: Added.
(WebKit::WebCookieCache::inMemoryStorageSession):

Canonical link: https://commits.webkit.org/289693@main
